### ISSUE TITLE
Python support?

### ISSUE THREAD

**Issue by twoATOMS** [TASVideos/BizHawk#1243]
Hello guys, I have a question about the emulator. I am trying to learn Machine Learning and want to apply it to video games, however I only know Python2.7 as a programming language and I was wondering if BizHawk will ever have Python scripting support in the near future?

Thanks.

**Comment by vadosnaprimer** [comment]
Clearly not in the "near future". Dunno about "ever" - maybe?

**Comment by spiiin** [comment]
@twoATOMS You might want to check out Nintaco (http://nintaco.com/api.html). It provides programmatic control using C, C#, Java, Lua and Python.

**Comment by g0me3** [comment]
maybe first you may try Man Learning? maybe learn lua or something to begin with learning the machines lol

**Comment by YoshiRulz** [comment]
see also previous rejection of Python #707

**Comment by vadosnaprimer** [comment]
Right, recently we had a talk, and if someone makes a PR, we'll most likely accept it.

To actually throw something on the table, here's the approach that looks very promising:
http://tasvideos.org/Feos/CallingCSharpMethodsFromPython.html
You expose the methods from C#, have a C# Python interpreter, and then Python scripts just call those methods dynamically.

**Comment by spiiin** [comment]
It's not very hard to implement lua server, and connect to it from ANY language via sockets.
Example for fceux emulator:
https://github.com/spiiin/fceux_luaserver

**Comment by vadosnaprimer** [comment]
Yeah, this dynamic approach should also allow any language with C# bindings to use those functions.

**Comment by elcolie** [comment]
+1 on Python websocket support for Machine Learning. Then with `Dockerfile` will be great!
I will make an article for building this. Can anyone here guide me through this?

**Comment by YoshiRulz** [comment]
Several people have suggested an ML project using BizHawk, so I'm writing this out to save repeating myself:
The easiest way to get memory access and playback controls is with an external tool. You could use Lua to bridge to another process, but it'll be slower. External tools need to be in C# (or, maybe, F# or VB). The library you'll use will probably be [ML.NET](https://dot.net/ml) what with Microsoft throwing money at it.
For completeness, you could use C# instead of Lua to bridge to another process. ~~I imagine that even calling into C from C# would be slower than a pure C# solution.~~ edit: Less sure about this, especially if you're running under Mono.

I'm personally willing to help anyone with the implementation of an external tool (I don't know much about ML, sorry). So, @elcolie, if you know that it's going to be annoying and still want to try, ~~I'm on IRC and [Matrix](https://matrix.to/#/@YoshiRulz:matrix.org)~~ [join the ApiHawk Discord](https://github.com/TASEmulators/BizHawk-ExternalTools/wiki).

**Comment by elcolie** [comment]
Thank you @YoshiRulz I will learn the material and when I am ready I will contact you 🙏

**Comment by magnusjjj** [comment]
Time to be annoying and necro a thread: https://github.com/magnusjjj/FakeControllerDesu

Rudimentary python support, at least for the restart function ;). I am looking for testers, to make sure that it's reasonably stable for other people to, well, I guess toy with.
